### PR TITLE
Add interface to mpi_gatherv

### PR DESCRIPTION
**Pull request by pecchia**

### CODE REVIEW

**Comment by bhourahine** [comment]
- Could you add a test example in mpifx/tests?
- Comments in mpifx_gather.m4 should explain the difference in receiving and sending process' displs and recvtype (not referenced in either case for receive?).

Also suggest adding wrapers that (analogous to the mpifx_gather subroutine_mpifx_gather_dr1) allow user set values of the buffer sizes and displs.

**Comment by aradi** [comment]
I think, an optional ``displ`` parameter would be indeed very useful, in case, the collected buffers should not be alligned continuously in the receive buffer.

However, I am actually wondering, whether it makes sense at all to provide wrappers for anything else than

* rank 0 buffers gathered into a rank 1 container (scalar -> vector)
* rank 1 buffers collected into a rank 1 container (vector -> vector)

Since the collected buffers have user specified sizes, one may end up with sizes which are not compatible with the leading dimension(s) of the multidimensional receiving buffer. I think, that would create quite a mess. What about only providing rank0 -> rank1 and rank1 -> rank1 wrappers for the moment?

Alternatively we require that the sending buffers and the receive buffers are shape compatible and the displ/receivcounts are given not in the units of a single element, but in the units of the smallest block meaningful for the given sending buffer shapes. (in case of the two cases above, that would boil down to single element units, but for higher rank arrays it would mean bigger blocks.)

**Comment by pecchia** [comment]
Ok, I added optional displs and, as discussed, the interface to scalar -> rank1 array. 
A test was added to check different cases

File: test/test_gatherv.f90 [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    allocate(recv1(nrecv))
+    allocate(recvcounts(mycomm%size))
+    recvcounts = 1
+    allocate(displs(mycomm%size))
+    ! set a non trivial displs vector
+    do ii = 1, mycomm%size 
+      displs(ii) = mycomm%size - ii 
+    end do
+  else
+    send0 = mycomm%rank + 1
```
Comment (bhourahine) [inline test/test_gatherv.f90]: Any reason my this is done inside the if block, unlike the two I? -> I? cases above (send0 could be uninitialized on the master, depending on compiler).

File: test/test_gatherv.f90 [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  else
+    allocate(recv1(0))
+    allocate(recvcounts(0))
+  end if
+
+  write(*, *) 'id:',mycomm%rank, "Send1 buffer:", send1(:)
+
+  call mpifx_gatherv(mycomm, send1, recv1, recvcounts)
+  
+  call mpifx_barrier(mycomm)
```
Comment (bhourahine) [inline test/test_gatherv.f90]: Are the barrier statements needed here? The gatherv call is blocking (https://stackoverflow.com/questions/9284419/is-mpi-reduce-blocking-or-a-natural-barrier) and at least on a shared memory machine this example isn't fast enough to cause problems (test_gather doesn't use these).

File: test/test_gatherv.f90 [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  integer, allocatable :: recvcounts(:)
+  integer, allocatable :: displs(:)
+  integer :: ii, nrecv  
+  character(100) :: formstr
+  character(*), parameter :: label = "(I2.2,'-',I3.3,'|',1X"
+
+  call mpifx_init()
+  call mycomm%init()
+
+  ! I1 -> I1
```
Comment (bhourahine) [inline test/test_gatherv.f90]: R1 -> R1 ?

File: test/test_gatherv.f90 [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+  if (mycomm%master) then
+    write(*, *) 'id:',mycomm%rank, "Recv1 buffer:", recv1
+    deallocate(recvcounts)
+    deallocate(recv1)
+  end if
+   
+  call mpifx_barrier(mycomm)
+
+  ! I2 -> I2
```
Comment (bhourahine) [inline test/test_gatherv.f90]: Real

File: test/test_gatherv.f90 [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  call mpifx_barrier(mycomm)
+
+  if (mycomm%master) then
+    write(*, *) "id:",mycomm%rank, "Recv2 buffer:", recv2(:,:)
+    deallocate(recvcounts)
+  end if
+
+  call mpifx_barrier(mycomm)
+
+  ! I0 -> I1
```
Comment (bhourahine) [inline test/test_gatherv.f90]: Real

File: test/test_gatherv.f90 [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    deallocate(recv1)
+  end if
+   
+  call mpifx_barrier(mycomm)
+
+  ! I2 -> I2
+  if (mycomm%master) write(*, *) 'Test gather rank=2 -> rank=2'
+
+  allocate(send2(10, mycomm%rank+1)) 
+  send2 = 1.0 * mycomm%rank
```
Comment (bhourahine) [inline test/test_gatherv.f90]: rank+1 used in the other test cases

File: src/mpifx_gatherv.F90 [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+  !!       real, allocatable :: recv1(:)
+  !!       integer, allocatable :: recvcounts(:)
+  !!       integer :: ii, nrecv  
+  !!       character(100) :: formstr
+  !!       character(*), parameter :: label = "(I2.2,'-',I3.3,'|',1X"
+  !!     
+  !!       call mpifx_init()
+  !!       call mycomm%init()
+  !!     
+  !!       ! I1 -> I1
```
Comment (bhourahine) [inline src/mpifx_gatherv.F90]: R1 as real ?

**Comment by pecchia** [comment]
I cleaned the test. The mpi_barriers where used just to obtain a nice screen output showing all send before recv.

**Comment by pecchia** [comment]
I checked out test_gatherv from Ben. I am fine without mpi_barries but in order to see an understandable output I removed the printout of the sent data. This because on my machines I get the output of the three examples completly interleaved in a random way. Now only master writes to screen.

**Comment by bhourahine** [comment]
All fine, now merged to main.

File: src/mpifx_gatherv.m4 [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  integer, intent(out), optional :: error
+
+  integer :: root0, error0, ii
+  integer, allocatable :: displs0(:)
+
+
+  _handle_inoptflag(root0, root, mycomm%masterrank)
+
+  if (mycomm%rank == root0) then
+    _assert(size(recv) == sum(recvcounts))
```
Comment (aradi) [inline src/mpifx_gatherv.m4]: That could go wrong, IMHO. If we use an optional displacement in order to fill up the receive buffer unevenly (with 'holes' between the chunks received from various processes), this condition won't be fulfilled.

File: src/mpifx_gatherv.m4 [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  integer, intent(in), optional :: root
+  integer, intent(out), optional :: error
+
+  integer :: ii, root0, error0
+  integer, allocatable :: displs0(:)
+
+  _handle_inoptflag(root0, root, mycomm%masterrank)
+
+  if (mycomm%rank == root0) then
+    _assert(size(recv) == sum(recvcounts))
```
Comment (aradi) [inline src/mpifx_gatherv.m4]: The same comment as above

File: src/mpifx_gatherv.m4 [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    _assert(size(recv, dim=$6) == mycomm%size)
```
Comment (aradi) [inline src/mpifx_gatherv.m4]: I think, that this also collides with the optional ``displs`` parameter.

Comment (pecchia) [reply src/mpifx_gatherv.m4]: The dr1- form as scalar to vector has a functionality similar to mpi_gather but the scalar can be arranged with an arbitrary permutation using displs. 
I did not think at the  possilility to leave holes in the gathered vector. This requires a more elaborate assert

**Pull request merged**